### PR TITLE
Changes for generating full static site

### LINKED ISSUE #308: nuxt build halts the machine
The app is currently running on a GCE g1-small (0.5vCPU & 1.7GB memory). `docker-compose up -d --build` is used for deploying, but it started to occasionally halt the entire machine in the past month. Cannot be reproduced in the staging environment.

- We might need to make sure we did not create any change that leads to memory leakage first. 
- Should also check if it's an issue regarding docker settings.
- One of the possible reasons is the building process takes up a significant amount of memory but g1-small only has 1.7 Gb. Due to the limited budget, limiting the memory usage of the building process would be a better option than upgrading the machine type.

### LINKED ISSUE #339: 2022 轉換靜態頁

### PR DESCRIPTION
**Pull request by josix**
- build(nuxt-generate): add dynamic routes configuration and update to use asyncData fetching data

## Types of changes

* **New feature**

## Description
- Update configuration to be able to generate dynamic routes
- Update to use asycData so that the pages could be full static. (I'm still figuring out the reason; the data only showed up in full static mode in this approach anyway.)

## Steps to Test This Pull Request
1. `npm run json-server`
2. `npm run generate`
3. `npm run start`
4. Check out if all pages works well.

## Expected behavior
- All pages should show up data as usual.

## Related Issue
(Not sure if it could fix_ #308)
Fix #339

### CODE REVIEW

**Comment by mattwang44** [comment]
CI failed due to no dependency of axios. (we use `npm ci` so you need to check-in the package-lock.json)

File: pages/conference/_eventType/_id.vue [hunk 1]
```diff
@@ -184,6 +182,19 @@ export default {
...
+        if (payload && Object.keys(payload).length !== 0) {
+            return {
+                speechData: payload,
+            }
+        }
+        await store.dispatch('$getSpeechData', {
+            eventType: params.eventType,
+            eventId: params.id,
+        })
+        const speechData = store.state.speechData
```
Comment (yungshenglu) [inline pages/conference/_eventType/_id.vue:195]: `store.state?.speechData`

File: pages/conference/_eventType/_id.vue [hunk 2]
```diff
@@ -184,6 +182,19 @@ export default {
...
     },
+    async asyncData({ store, params, payload }) {
+        if (payload && Object.keys(payload).length !== 0) {
+            return {
+                speechData: payload,
+            }
+        }
+        await store.dispatch('$getSpeechData', {
+            eventType: params.eventType,
+            eventId: params.id,
```
Comment (yungshenglu) [inline pages/conference/_eventType/_id.vue:193]: eventType: params?.eventType,
eventId: params?.id,

File: pages/conference/_eventType/index.vue [hunk 3]
```diff
@@ -73,16 +73,23 @@ export default {
...
-    asyncData({ redirect, params }) {
+    async asyncData({ redirect, params, store }) {
         const eventType = params.eventType
         if (!['talks', 'tutorials'].includes(eventType)) {
             redirect('/')
         }
+        await store.dispatch('$getSpeechesData', eventType)
+        const speechesData = store.state.speechesData.map((speech) => ({
+            ...speech,
+            begin_time: speech.begin_time ? new Date(speech.begin_time) : null,
```
Comment (yungshenglu) [inline pages/conference/_eventType/index.vue:84]: `speech?.begin_time`

File: pages/conference/keynotes.vue [hunk 4]
```diff
@@ -141,6 +141,17 @@ export default {
         GithubIcon,
         TwitterIcon,
     },
+    async asyncData({ store, app }) {
+        await store.dispatch('$getKeynotesData')
+
+        const keynotesData = store.state.keynotesData.map((keynote) => ({
```
Comment (yungshenglu) [inline pages/conference/keynotes.vue:147]: `store.state?.keynotesData.map....`

File: pages/index.vue [hunk 5]
```diff
@@ -120,6 +119,13 @@ export default {
         I18nPageWrapper,
         Intro,
     },
+    async asyncData({ store }) {
+        await store.dispatch('$getSponsorsData')
+        const sponsorsData = store.state.sponsorsData
```
Comment (yungshenglu) [inline pages/index.vue:124]:
```js
const sponsorsData = store.state?.sponsorsData
```

File: pages/events/jobs.vue [hunk 6]
```diff
@@ -39,19 +39,24 @@ export default {
         JobsCardCollection,
         JobsPanel,
     },
+    async asyncData({ store, app }) {
+        await store.dispatch('$getJobsData')
+        const jobsData = store.state.jobsData.map((sponsor) => ({
```
Comment (yungshenglu) [inline pages/events/jobs.vue:44]:
```js
const jobsData = store.state?.jobsData.map((sponsor) => ({
```

File: pages/conference/_eventType/index.vue [hunk 7]
```diff
@@ -73,16 +73,23 @@ export default {
...
         Banner,
     },
-    asyncData({ redirect, params }) {
+    async asyncData({ redirect, params, store }) {
         const eventType = params.eventType
         if (!['talks', 'tutorials'].includes(eventType)) {
             redirect('/')
         }
+        await store.dispatch('$getSpeechesData', eventType)
+        const speechesData = store.state.speechesData.map((speech) => ({
```
Comment (yungshenglu) [inline pages/conference/_eventType/index.vue:82]:
```js
const speechesData = store.state?.speechesData.map((speech) => ({
```

File: nuxt.config.js [hunk 8]
```diff
@@ -11,7 +13,64 @@ export default {
...
+            }
+            const getAllDetailTutorials = async () => {
+                const data = await Promise.all(
+                    tutorials.data.map(async (tutorial) => {
+                        return await axios
+                            .get(
+                                `${DEFAULT_BASE_URL}/api/events/speeches/${tutorial.event_type}/${tutorial.id}/`,
+                                config,
+                            )
+                            .then((response) => response.data)
```
Comment (yungshenglu) [inline nuxt.config.js:53]:
```js
.then((response) => response?.data)
```

**Comment by josix** [comment]
Duplicated with #380

**Pull request closed without merging**